rust/core-lib/src/editor.rs

```diff
@@ -712,6 +717,71 @@ impl Editor {
+    /// Uses a BTreeSet to holds the collection of lines to modify.
+    /// Preserves cursor position and current selection as much as possible.
+    /// Tries to have behavior consistent with other editors like Atom, Sublime and VSCode,
+    /// with non-caret selections not being modified.
+    fn modify_indent(&mut self, direction: IndentDirection) {
+        let mut lines = BTreeSet::new();
+        let tab_text = if self.config.items.translate_tabs_to_spaces {
+                let tab_size = self.config.items.tab_size;
+                n_spaces(tab_size)
+            } else { "\t" };
+        for region in self.view.sel_regions() {
+            let (first_line, _) = self.view.offset_to_line_col(&self.text, region.min());
+            let (last_line, last_col) =
+                self.view.offset_to_line_col(&self.text, region.max());
+            let last_line = if last_col == 0 && last_line > first_line {
+                last_line - 1
+            } else {
+                last_line   
+            };
+            let line_range = first_line..(last_line + 1);
+            for line in line_range {
+                lines.insert(line);
+            }
+        }    
+        match direction {
+            IndentDirection::In => {
+               self.indent(lines, tab_text);
+            },
+            IndentDirection::Out => {
+              self.outdent(lines, tab_text);
+            },
+        };
+    }
+
+    fn indent(&mut self, lines: BTreeSet<usize>, tab_text: &str) {
+        let mut builder = delta::Builder::new(self.text.len());
+        for line in lines {
+            let offset = self.view.line_col_to_offset(&self.text, line, 0);
+            let interval = Interval::new_closed_open(offset, offset);  
+            builder.replace(interval, Rope::from(tab_text));
+            self.this_edit_type = EditType::InsertChars;
+        }      
+        self.add_delta(builder.build());   
+    }
+
+    fn outdent(&mut self, lines: BTreeSet<usize>, tab_text: &str) {
+        let mut builder = delta::Builder::new(self.text.len());
+        for line in lines {
+            let offset = self.view.line_col_to_offset(&self.text, line, 0);
+            let tab_offset = self.view.line_col_to_offset(&self.text, line, tab_text.len());
+            let interval = Interval::new_closed_open(offset, tab_offset);         
+            let leading_slice = self.text.slice_to_string(interval.start(), interval.end());
+            if leading_slice == tab_text {
+                builder.delete(interval);
+            } else if let Some(first_char_col) = leading_slice.find(|c: char| !c.is_whitespace()) {
+                let first_char_offset = self.view.line_col_to_offset(&self.text, line, first_char_col);
+                let interval = Interval::new_closed_open(offset, first_char_offset);
+                builder.delete(interval);
+            }
+            self.this_edit_type = EditType::Delete;
```

Comment:
we only need to set this once, not on each iteration of the loop. (same as up top)

Reply:
I’m not sure what you mean here, do you mean the variables in the loop? They seem dependent on the line to me.

Reply:
I mean you set `self.this_edit_type` for each line you adjust, when you only need to set it once for the whole operation.